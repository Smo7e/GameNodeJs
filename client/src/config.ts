// UdSU
// const DOMAIN = 'http://studfront';
// const PORT = 81;

// dev
const DOMAIN = "http://localhost";
const PORT = 3001;

// prod
//...

export const HOST = PORT ? `${DOMAIN}:${PORT}` : `${DOMAIN}`;

export const MEDIATOR = {
    EVENTS: {
        SERVER_ERROR: "SERVER_ERROR",
        GET_MESSAGES: "GET_MESSAGES",
        GET_SCENE: "GET_SCENE",
        GET_USER: "GET_USER",
        GET_GAMERS: "GET_GAMERS",
        GET_FRIENDS: "GET_FRIENDS",
        GET_INVITES: "GET_INVITES",
        LOGIN: "LOGIN",
        SIGNUP: "SIGNUP",
        LOGOUT: "LOGOUT",
        GET_MOBS: "GET_MOBS",
        GET_ERROR: "GET_ERROR",
        PLAY_MUSIC: "PLAY_MUSIC", 
        STOP_MUSIC: "STOP_MUSIC", 
        SET_MUSIC_VOLUME: "SET_MUSIC_VOLUME", 
        UPDATE_ARR_BULLET_TRAJECTORY: "UPDATE_ARR_BULLET_TRAJECTORY",
    },
    TRIGGERS: {},
};
