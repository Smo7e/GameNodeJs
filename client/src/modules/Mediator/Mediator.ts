type TEVENT = {
    [key: string]: string;
};

type TParams = {
    EVENTS: TEVENT;
    TRIGGERS: TEVENT;
};

type TFunction = <T>(data: any) => T | void | null;

type TEvents = {
    [key: string]: Array<TFunction>;
};

type TTriggers = {
    [key: string]: TFunction;
};

export default class Mediator {
    private events: TEvents;
    private EVENTS: TEVENT;
    private triggers: TTriggers;
    private TRIGGERS: TEVENT;

    constructor(params: TParams) {
        const { EVENTS, TRIGGERS } = params;
        this.EVENTS = EVENTS;
        this.TRIGGERS = TRIGGERS;
        this.events = {};
        Object.keys(EVENTS).forEach((key) => (this.events[key] = []));
        this.triggers = {};
        Object.keys(TRIGGERS).forEach(
            (key) =>
                (this.triggers[key] = () => {
                    return null;
                })
        );
    }

    /***************/
    /* Про события */
    /***************/
    getEventTypes() {
        return this.EVENTS;
    }

    subscribe(name: string, func: TFunction): void {
        if (name && this.events[name] && func instanceof Function) {
            this.events[name].push(func);
        }
    }

    unsubscribe(name: string, func: TFunction): void {
        if (name && this.events[name] && func instanceof Function) {
            this.events[name] = this.events[name].filter((handler) => handler !== func);
        }
    }

    call<T>(name: string, data: T | null = null): void {
        if (name && this.events[name]) {
            this.events[name].forEach((func) => func instanceof Function && func(data));
        }
    }

    /****************/
    /* Про триггеры */
    /****************/
    getTriggerTypes() {
        return this.TRIGGERS;
    }

    set(name: string, func: TFunction): void {
        if (name && this.triggers[name] && func instanceof Function) {
            this.triggers[name] = func;
        }
    }

    get<T>(name: string, data: any = null): T | null {
        if (name && this.triggers[name] && this.triggers[name] instanceof Function) {
            return this.triggers[name](data) as T;
        }
        return null;
    }
}
