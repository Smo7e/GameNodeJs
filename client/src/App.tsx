import React, { useState } from "react";
import { HOST, MEDIATOR } from "./config";
import { Server, Mediator, Store, TError } from "./modules";
import SignUp from "./component/SignUp/SignUp";
import Game from "./modules/Game/Game";
import Menu from "./component/Menu/Menu";
import Login from "./component/Login/Login";
import Heroes from "./component/Heroes/Heroes";
import Interface from "./component/Interface/Interface";
import Lobby from "./component/Lobby/Lobby";
import ErrorMessage from "./modules/ErrorMessage/ErrorMessage";
import Parameters from "./component/Parameters/Parameters";
import Music from "./component/Parameters/Music";
import "./App.css";

export const ServerContext = React.createContext<Server>(null!);
export const MediatorContext = React.createContext<Mediator>(null!);
export const StoreContext = React.createContext<Store>(null!);

export enum EPAGES {
    SIGNUP,
    LOGIN,
    GAME,
    MENU,
    HEROES,
    LOBBY,
    PARAMETERS,
}

const MainApp = () => {
    const [epages, setEpages] = useState<EPAGES>(EPAGES.LOGIN);
    return (
        <>
            {epages === EPAGES.LOGIN ? (
                <Login epages={setEpages} />
            ) : epages === EPAGES.SIGNUP ? (
                <SignUp epages={setEpages} />
            ) : epages === EPAGES.MENU ? (
                <Menu epages={setEpages} />
            ) : epages === EPAGES.GAME ? (
                <>
                    <Game />
                    <Interface />
                </>
            ) : epages === EPAGES.HEROES ? (
                <Heroes epages={setEpages} />
            ) : epages === EPAGES.LOBBY ? (
                <Lobby epages={setEpages} />
            ) : epages === EPAGES.PARAMETERS ? (
                <Parameters epages={setEpages} />
            ) : (
                <></>
            )}
            <ErrorMessage />
            <Music />
        </>
    );
};

const App: React.FC = () => {
    const store = new Store();
    const mediator = new Mediator(MEDIATOR);
    const server = new Server(HOST, mediator, store);

    const { SERVER_ERROR } = mediator.getEventTypes();
    mediator.subscribe(SERVER_ERROR, (data: TError) => {
        console.log(data);
    });

    return (
        <StoreContext.Provider value={store}>
            <MediatorContext.Provider value={mediator}>
                <ServerContext.Provider value={server}>
                    <MainApp />
                </ServerContext.Provider>
            </MediatorContext.Provider>
        </StoreContext.Provider>
    );
};

export default App;
