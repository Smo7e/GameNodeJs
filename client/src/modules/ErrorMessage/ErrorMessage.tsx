import React, { useState, useEffect, useContext } from "react";
import { TError } from "../Server/types";

import "./ErrorMessage.css";
import { MediatorContext } from "../../App";

const ErrorMessage: React.FC = () => {
    const mediator = useContext(MediatorContext);

    const [isVisible, setIsVisible] = useState(true);
    const [error, setError] = useState<TError | null>(null);

    useEffect(() => {
        const { SERVER_ERROR } = mediator.getEventTypes();
        const errorHandler = (data: TError) => {
            setIsVisible(true);
            setError(data);
        };

        mediator.subscribe(SERVER_ERROR, errorHandler);
        if (error !== null) {
            const timer = setTimeout(() => {
                setIsVisible(false);
            }, 10000);

            return () => {
                mediator.unsubscribe(SERVER_ERROR, errorHandler);
                clearTimeout(timer);
            };
        }
    });

    const handleClose = () => {
        setIsVisible(false);
    };

    const handleOk = () => {
        console.log("OK button clicked");
        setIsVisible(false);
    };
    if (error === null) {
        return <></>;
    }

    if (!isVisible) {
        return <></>;
    }

    return (
        <div className="error-container">
            <div className="error-header">
                <span className="error-title">Ошибка: {error?.code}</span>
                <button className="close-button" onClick={handleClose}></button>
            </div>
            <div className="error-body">
                <div className="error-text" id="test-error-text">
                    {error?.text}
                </div>
            </div>
            <div className="error-footer">
                <button className="ok-button" onClick={handleOk}>
                    OK
                </button>
            </div>
        </div>
    );
};

export default ErrorMessage;
