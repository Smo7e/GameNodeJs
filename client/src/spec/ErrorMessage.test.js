import React from 'react'; 
import { render } from '@testing-library/react'; 
import ErrorMessage from '../modules/ErrorMessage/ErrorMessage'; 
import { MediatorContext } from '../App'; 
import Mediator from '../modules/Mediator/Mediator'; 
import { HOST, MEDIATOR } from "../config";  
 
jest.useFakeTimers(); 
 
describe('ErrorMessage Component', () => { 
    const mediator = new Mediator(MEDIATOR);  
   
  it('renders correctly with null error', () => { 
    const { asFragment } = render( 
      <MediatorContext.Provider value={mediator}> 
        <ErrorMessage /> 
      </MediatorContext.Provider> 
    ); 
    expect(asFragment()).toMatchSnapshot(); 
  }); 
 
  it('renders correctly with error and is visible', () => { 
    const error = { code: 500, text: 'Internal Server Error' }; 
    mediator.call('SERVER_ERROR', error);  
     
    const { asFragment } = render( 
      <MediatorContext.Provider value={mediator}> 
        <ErrorMessage /> 
      </MediatorContext.Provider> 
    ); 
    expect(asFragment()).toMatchSnapshot(); 
  }); 
 
  it('hides after 10 seconds', () => { 
    const error = { code: 404, text: 'Not Found' }; 
    mediator.call('SERVER_ERROR', error);   
 
    const { asFragment } = render( 
      <MediatorContext.Provider value={mediator}> 
        <ErrorMessage /> 
      </MediatorContext.Provider> 
    ); 
 
    jest.advanceTimersByTime(10000); 
    expect(asFragment()).toMatchSnapshot();  
  }); 
});
