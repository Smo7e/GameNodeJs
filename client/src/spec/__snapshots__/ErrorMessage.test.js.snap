// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage Component hides after 10 seconds 1`] = `<DocumentFragment />`;

exports[`ErrorMessage Component renders correctly with error and is visible 1`] = `<DocumentFragment />`;

exports[`ErrorMessage Component renders correctly with null error 1`] = `<DocumentFragment />`;
